### PR TITLE
[CDAP-18384] Authorization Metrics

### PR DESCRIPTION
**Pull request by dli357**
Added the following metrics for authorization:
 - `authorization.internal.check.success.count` - Successful internal auth checks (counter)
 - `authorization.internal.check.failure.count` - Failed internal auth checks (counter)
 - `authorization.internal.visibility.check.count` - Internal visibility checks (counter)
 - `authorization.extension.check.success.count` - Successful extension checks (counter)
 - `authorization.extension.check.failure.count` - Failed extension checks (counter)
 - `authorization.extension.check.bypass.count` - Bypassed extension checks (due to master principal) (counter)
 - `authorization.non.internal.visibility.check.count` - Non-internal visibility checks (counter)
 - `authorization.extension.check.millis` - Milliseconds taken per authorization extension check (gauge)
 - `authorization.extension.visibility.millis` - Milliseconds taken per authorization extension visibility check (gauge)

### CODE REVIEW

File: cdap-security/src/main/java/io/cdap/cdap/security/authorization/DefaultAccessEnforcer.java [hunk 1]
```diff
@@ -226,4 +275,66 @@ private boolean isEnforcingOnSamePrincipalId(EntityId entityId, Principal princi
     return entityId.getEntityType().equals(EntityType.KERBEROSPRINCIPAL) &&
       principal.getName().equals(entityId.getEntityName());
   }
+
+  /**
+   * Constructs metrics tags for a given entity ID.
+   */
+  static Map<String, String> createEntityIdMetricsTags(EntityId entityId) {
+    Map<String, String> tags = new HashMap<>();
+    EntityId currEntityId = entityId;
```
Comment (tivv) [inline cdap-security/src/main/java/io/cdap/cdap/security/authorization/DefaultAccessEnforcer.java]: You can simply do `for (EntityId currentEntity: enityId.getHierarchy())`

File: cdap-security/src/test/java/io/cdap/cdap/security/authorization/DefaultAccessEnforcerTest.java [hunk 2]
```diff
@@ -276,7 +286,7 @@ public void testIsVisibleWithEncryptedCredential()
            new AccessControllerInstantiator(CCONF, AUTH_CONTEXT_FACTORY)) {
       AccessController accessController = accessControllerInstantiator.get();
       DefaultAccessEnforcer accessEnforcer = new DefaultAccessEnforcer(CCONF, sConfCopy, accessControllerInstantiator,
```
Comment (tivv) [inline cdap-security/src/test/java/io/cdap/cdap/security/authorization/DefaultAccessEnforcerTest.java]: nit: move to setup / factory method. This way if we need to change constructor we don't need to update in 10 places

File: cdap-security/src/test/java/io/cdap/cdap/security/authorization/DefaultAccessEnforcerTest.java [hunk 3]
```diff
@@ -256,7 +266,7 @@ public void testAuthEnforceWithBadEncryptedCredential()
                              ImmutableSet.of(StandardPermission.GET, StandardPermission.GET));
 
       DefaultAccessEnforcer accessEnforcer = new DefaultAccessEnforcer(CCONF, sConfCopy, accessControllerInstantiator,
-                                                                       null);
+                                                                       null, new NoOpMetricsCollectionService());
```
Comment (tivv) [inline cdap-security/src/test/java/io/cdap/cdap/security/authorization/DefaultAccessEnforcerTest.java]: Do we test anywhere that metrics calls are actually done? It's better to add mocks and assert calls are done

File: cdap-security/src/main/java/io/cdap/cdap/security/authorization/DefaultAccessEnforcer.java (hunk 1 above)
Comment (dli357) [reply cdap-security/src/main/java/io/cdap/cdap/security/authorization/DefaultAccessEnforcer.java]: Fixed

File: cdap-security/src/test/java/io/cdap/cdap/security/authorization/DefaultAccessEnforcerTest.java (hunk 2 above)
Comment (dli357) [reply cdap-security/src/test/java/io/cdap/cdap/security/authorization/DefaultAccessEnforcerTest.java]: Fixed

File: cdap-security/src/test/java/io/cdap/cdap/security/authorization/DefaultAccessEnforcerTest.java (hunk 3 above)
Comment (dli357) [reply cdap-security/src/test/java/io/cdap/cdap/security/authorization/DefaultAccessEnforcerTest.java]: Fixed

**Comment by dli357** [comment]
Thanks for review!

**Review by tivv — APPROVED** [review]

File: cdap-security/src/main/java/io/cdap/cdap/security/authorization/DefaultAccessEnforcer.java [hunk 4]
```diff
@@ -150,16 +191,22 @@ public void enforceOnParent(EntityType entityType, EntityId parentId, Principal
...
   public Set<? extends EntityId> isVisible(Set<? extends EntityId> entityIds, Principal principal)
     throws AccessException {
+    // Pass null for creating metrics context. Aggregations are not supported for visibility checks.
+    MetricsContext metricsContext = createEntityIdMetricsContext(null);
     if (internalAuthEnabled && principal.getFullCredential() != null
       && principal.getFullCredential().getType() == Credential.CredentialType.INTERNAL) {
       LOG.trace("Internal Principal enforce({}, {})", entityIds, principal);
-      return internalAccessEnforcer.isVisible(entityIds, principal);
+      metricsContext.increment(Constants.Metrics.Authorization.INTERNAL_VISIBILITY_CHECK_COUNT, 1);
+      Set<? extends EntityId> visibleEntities = internalAccessEnforcer.isVisible(entityIds, principal);
```
Comment (chtyim) [inline cdap-security/src/main/java/io/cdap/cdap/security/authorization/DefaultAccessEnforcer.java]: You can just return without introducing a local variable.

File: cdap-security/src/main/java/io/cdap/cdap/security/authorization/DefaultAccessEnforcer.java [hunk 5]
```diff
@@ -226,4 +274,63 @@ private boolean isEnforcingOnSamePrincipalId(EntityId entityId, Principal princi
...
+   * Constructs tags and returns a metrics context for a given entity ID.
+   */
+  private MetricsContext createEntityIdMetricsContext(EntityId entityId) {
+    if (!metricsCollectionEnabled) {
+      return new NoopMetricsContext();
+    }
+    Map<String, String> tags;
+    if (metricsTagsEnabled && entityId != null) {
+      tags = createEntityIdMetricsTags(entityId);
+    } else {
```
Comment (chtyim) [inline cdap-security/src/main/java/io/cdap/cdap/security/authorization/DefaultAccessEnforcer.java]: Usually I prefer fewer branching as the code would be cleaner.

```
Map<String, String> tags = Collections.emptyMap();
if (...) {
  tags = createEntityIdMetricsTags(entityId);
}
```

**Review by chtyim — APPROVED** [review]

File: cdap-security/src/main/java/io/cdap/cdap/security/authorization/DefaultAccessEnforcer.java (hunk 4 above)
Comment (dli357) [reply cdap-security/src/main/java/io/cdap/cdap/security/authorization/DefaultAccessEnforcer.java]: Fixed

File: cdap-security/src/main/java/io/cdap/cdap/security/authorization/DefaultAccessEnforcer.java (hunk 5 above)
Comment (dli357) [reply cdap-security/src/main/java/io/cdap/cdap/security/authorization/DefaultAccessEnforcer.java]: Fixed

**Pull request merged**